### ISSUE TITLE
Failure on type application

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/hlint#967]
```
loadCradleOnlyonce :: TestTree
loadCradleOnlyonce = skipManyTill anyMessage (message @PublishDiagnosticsNotification)
```

Gives:

```
Sample.hs:2:46-86: Warning: Redundant bracket
Found:
  (_)
Perhaps:
  _
```

That's a terrible suggestion because the bracket is required and it seems to have got super confused about what is there. Originally from https://github.com/digital-asset/ghcide/pull/544 by @cocreature

**Comment by ndmitchell** [comment]
If TypeApplications is turned off (as we default to in HLint) then `skipManyTill anyMessage (message @PublishDiagnosticsNotification)` is parsed as `skipManyTill anyMessage (_)`. If I run it through GHC without that extension then we get a syntax error. It looks like we must be forgetting an error message somewhere? (cc @shayne-fletcher - it seems like its in the pattern vs expression thing you blogged about - any insights?)

However, the reason for not enabling TypeApplication was because the HSE parser tripped over things pretty badly. We can now enable it unconditionally.

**Comment by ndmitchell** [comment]
@shayne-fletcher - concretely we call parseFile in ghc-lib-parser-ex. I seem to remember that at some point merely getting a successful parse from GHC wasn't enough to ensure that you had a correct parse tree and you had to look elsewhere to get an error too. Or am I misremembering?

**Comment by shayne-fletcher** [comment]
No, you are not misremembering. Details in http://blog.shaynefletcher.org/2019/05/handling-ghc-parser-errors-right.html.

**Comment by shayne-fletcher** [comment]
The long and short of is that `POk` means "I could muddle through and get a parse tree". One should then consult if there are any errors in the parse state though to determine if the tree is good.

**Comment by shayne-fletcher** [comment]
The section of code where we'd check (but don't currently) is in `HSE/All.hs`:
```
            case fileToModule file ppstr ghcFlags of
                POk pst a ->
                    let anns =
                          ( Map.fromListWith (++) $ annotations pst
                          , Map.fromList ((noSrcSpan, comment_q pst) : annotations_comments pst)
                          ) in
                    pure $ Right (ModuleEx (applyFixities fixities a) anns)
```

**Comment by shayne-fletcher** [comment]
Your analysis seems very plausible:
```haskell
  mkHsAsPatPV l v e = do
    opt_TypeApplications <- getBit TypeApplicationsBit
    let msg | opt_TypeApplications
            = "Type application syntax requires a space before '@'"
            | otherwise
            = "Did you mean to enable TypeApplications?"
    patSynErr l (pprPrefixOcc (unLoc v) <> text "@" <> ppr e) (text msg)
```

**Comment by shayne-fletcher** [comment]
Is that the parse error that you see with the extension disabled?

**Comment by shayne-fletcher** [comment]
Must be. Indeed, it is classified as a non-fatal error and a hole expression is returned.
```haskell
patSynErr :: SrcSpan -> SDoc -> SDoc -> PV (LHsExpr GhcPs)
patSynErr l e explanation =
  do { addError l $
        sep [text "Pattern syntax in expression context:",
             nest 4 (ppr e)] $$
        explanation
     ; return (cL l hsHoleExpr) }

hsHoleExpr :: HsExpr (GhcPass id)
hsHoleExpr = HsUnboundVar noExtField (TrueExprHole (mkVarOcc "_"))
```

**Comment by ndmitchell** [comment]
Indeed, that's the error. I think we previously concluded that a muddled through parse tree was just fine and dandy. But given that a muddled through parse tree has complete garbage in it, failing to parse is definitely the preferable option. Sounds like a check in `fileToModule` would be a very good idea.

**Comment by shayne-fletcher** [comment]
> Indeed, that's the error. I think we previously concluded that a muddled through parse tree was just fine and dandy.

That's exactly how I remember it too.

> But given that a muddled through parse tree has complete garbage in it, failing to parse is definitely the preferable option. Sounds like a check in `fileToModule` would be a very good idea.

I am happy to provide a patch. Do you mind though if we do it at the level of `parseModuleEx`?

**Comment by ndmitchell** [comment]
Patch would be most welcome. `parseModuleEx` looks like a great spot to do it.

**Comment by shayne-fletcher** [comment]
PR here https://github.com/ndmitchell/hlint/pull/968

**Comment by ndmitchell** [comment]
So we've taken a double pronged approach. Enable type application by default and fail properly if there is a parse error, both of which fix this bug.

**Comment by ndmitchell** [comment]
Released as 3.0.4.
